### PR TITLE
Feature: logger state dumping and resuming

### PR DESCRIPTION
**Pull request by ashleve**
## What does this PR do?

Implements state dumping and resuming mechanism for loggers.
Implements `state_dict` and `load_state_dict()` for tensorboard.

Related to #5342

This PR is a reimplementation of https://github.com/PyTorchLightning/pytorch-lightning/pull/6361

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by ashleve** [comment]
Currently resuming tensorboard resets the logged step number https://github.com/PyTorchLightning/pytorch-lightning/issues/12274

File: tests/tests_pytorch/checkpointing/test_model_checkpoint.py [hunk 1]
```diff
@@ -870,17 +870,19 @@ def validation_step(self, batch, batch_idx):
         "enable_progress_bar": False,
         "enable_model_summary": False,
     }
-    trainer = Trainer(**trainer_kwargs, callbacks=[checkpoint_callback])
+    trainer = Trainer(**trainer_kwargs, default_root_dir=tmpdir, callbacks=[checkpoint_callback])
```
Comment (ashleve) [inline tests/tests_pytorch/checkpointing/test_model_checkpoint.py:873]: without this change tensorboard would initialize its version by incrementing last version number from other tests (should be init as `version_0` for assert below)

**Comment by ashleve** [comment]
@carmocca Thank you for the help! I fixed the tests accordingly

**Comment by ashleve** [comment]
@awaelchli @carmocca Please review the changes when you find the time

**Comment by Borda** [comment]
@carmocca, how is it going here? :otter:

**Comment by carmocca** [comment]
@awaelchli It would be nice to finish it. Since you are familiar with the state key, do you want to take it? Otherwise I will. Not a high priority of course.

**Comment by Borda** [comment]
@ashleve still working on this one? 🐰
